text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,1232 @@
+    moduleType?: "HTML" | "CSS" | "Javascript";
+
+    defaultComponentClass?: ComponentClass[];
+
+    templates?: { [key: string]: HTMLTemplateFile[]};
+
+    scripts?: { [key: string]: ScriptFile[]};
+
+    css?: { [key: string]: CSSFile[]};
+}
+```
+
+#### HTML Template file metadata shape
+```ts
+// Root metadata object for an HTML template file
+interface HTMLTemplateFile {
+    fileType: 'html';
+    fileName: string;
+    // List of unique components referenced in the template
+    componentReferences?: ComponentReference[];
+    // Static resources referenced in a template. For example: 'src' attribute value of an <img>, <source> tag.
+    staticResources?: StaticResourceReference[];
+    // Template in AST format starting with the root <template> node. Note: this is a partial AST
+    astRoot?: Node;
+}
+
+// Represents a component referenced in the template.
+// Use the 'tagName' property for kebab case and 'name' property for the reference in canonical form
+interface ComponentReference extends ModuleReference {
+    // Component reference always has a namespace
+    namespace: string;
+    // No moduleIdentifier for a component
+    moduleIdentifier?: never;
+    // Components are always imported from an external bundle
+    type: 'external';
+    tagName: string;
+}
+
+// Information about a node in the html template
+interface Node {
+    tagName: string;
+    location: SourceLocation;
+    // Any properties set in the template, will include html attributes, public properties
+    attributes?: ElementAttributeReference[];
+    // Event listeners attached declaratively in the template with a 'on' prefix, eg: 'onclick'
+    eventListeners?: DeclarativeEventListener[];
+    children?: Node[];
+}
+
+// Information about a custom element referenced in a HTML template.
+interface CustomElementNode extends Node {
+    // Information about slots set by parent
+    slots?: SlotReference[];
+}
+
+/**
+ * Information about default and named slots being set by the parent component. The parent can
+ * specify a 'slot' attribute in the slotted content in the case of named slot. Slot content that
+ * do not have a 'slot' attribute will be passed as default slot content. The 'name' property in
+ * the SlotReference object will be 'default' for such a case.
+ */
+interface SlotReference {
+    name: string;
+}
+
+// Information about an element referenced in the template with a 'lwc:dynamic' attribute.
+interface DynamicCustomElementNode extends CustomElementNode {
+    ctor: ComputedValue;
+}
+
+// Information about for:each directive usage to render a list in a html template.
+interface TemplateForEachDirective extends Node {
+    tagName: 'template';
+    directive: 'for:each';
+    items: ComputedValue;
+    itemName?: string;
+    indexName?: string;
+    key?: string;
+}
+
+// Information about if:true or if:false directive usage to perform conditional rendering in a html template.
+interface TemplateIfDirective extends Node {
```

Comment:
do we want these explicitly typed? We could introduce TemplateDirective { tagName, directiveKey, directiveValue }